src/differentials/composite.jl

```diff
@@ -49,7 +49,7 @@ function Base.:(==)(a::Composite{P, T}, b::Composite{P, T}) where {P, T}
     return backing(a) == backing(b)
 end
 function Base.:(==)(a::Composite{P}, b::Composite{P}) where {P, T}
-    return canonicalize(a) == canonicalize(b)
+    return extern(canonicalize(a)) == extern(canonicalize(b))
```

Comment:
We don't want this see #56, we want to get rid  of `extern`.
(and with the changes in ChainRulesTestUtils we are much closer than before).

We also want to remove the `canonicalize` from here probably.
So we can compare `Composites` that do not match the fields of their structs.

We might actually want the opposite of `canonicalize`: `simplify` that just drops any fields that are `::AbstractZero`.
but for this PR we probably want to union the keys of each backing and then check all are equal.
Probably just with a loop over the union of keys.

Reply:
I don't understand why we should be removing `canonicalize`?

If we are comparing two `Composite`s, one with explicit and one with implicit `Zero()`, we need either `canonicalize` or `simplify`, right? Otherwise the loop over the `Union` of keys will get `BoundsError`s or `NamedTuple has no field x`?

Reply:
> Otherwise the loop over the Union of keys will get BoundsErrors or NamedTuple has no field x?

No, because for any field it doesn't have the `Composite` returns `Zero()` on `getproperty`.

Plus there are many ways things can be equal without having the same backing representation.
One is thunks (`extern` fixes), one is having fields not present (`canonicalize` fixes)
But another is different floating point types (`Float32` and `Float64`) or having different matrix types (`FillArray` vs normal Dense `Matrix` vs `StaticArray`).
Some of these arguably are their own problems.

Also `canoncialize` fails if your derivatives are being stored in a way that didn't match to the fields of the primal.
Which we do do for Factorizations like `Cholesky`, we instead store them to match the properties of the primal, because the math is easier (though we have talked about changing it a few times).

Reply:
Makes sense, thanks!

Reply:
Should we add a test for Floats and matrix types?

Reply:
Might be a good idea, yes.
One or the other, doesn't have to be both.
Maybe using `1.0:3.0` vs  `[1.0, 2.0, 3.0]`

Reply:
I've restructured the NamedTuple example to include the Matrix already, it's not as elegant but should do the trick